File: Best Time to Buy and Sell Stock IV/mysolution3.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n=len(prices)
        if n<=1 or k==0:
            return 0
        
        # sell[i][j]: the max profit from day 0 to day i with j+1 transactions when the last transaction is sell;
        # buy[i][j]: the max profit from day 0 to day i with j+1 transactions when the last transaction is buy;
        sell=[[-inf for j in range(k)] for i in range(n)]
        buy=[[-inf for j in range(k)] for i in range(n)]
        buy[0][0]=-prices[0]
        for j in range(k):
            for i in range(1,n):
                if j==0:
                    buy[i][j]=max(buy[i-1][j],-prices[i])
                else:
                    buy[i][j]=max(buy[i-1][j],sell[i-1][j-1]-prices[i])
                
                sell[i][j]=max(sell[i-1][j],buy[i-1][j]+prices[i])
        
        ans=0
        for j in range(k):
            ans=max(ans,sell[n-1][j])
        
        return ans

File: Best Time to Buy and Sell Stock III/mysolution3.java
class Solution {
    public int maxProfit(int[] prices) {
        
        int n=prices.length;
        
        if(n<=1)
            return 0;
        
        int[][] S=new int[n][3];
        int[][] B=new int[n][3];
        
        for(int j=1;j<=2;j++)
        {
            S[0][j]=0;
            B[0][j]=-prices[0];
        }
        
        for(int i=1;i<n;i++)
        {
            for(int j=1;j<=2;j++)
            {
                S[i][j]=Math.max(S[i-1][j],B[i-1][j]+prices[i]);
                B[i][j]=Math.max(B[i-1][j],S[i-1][j-1]-prices[i]);
            }
        }
        
        return S[n-1][2];
    }
}